### ISSUE TITLE
Qualifies optional keys in object-like arrays

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#634]
This uses undocumented optional key syntax for object-like arrays.
```php
<?php
namespace N;

/**
 * @psalm-param array{key?:string} $p
 */
function f(array $p): void
{
    echo isset($p['key']) ? $p['key'] : "";
}
```
Fails with 
```
ERROR: InvalidArrayOffset - q.php:9:16 - Cannot access value on variable $p using offset value of 'key', expecting 'N\key'
    echo isset($p['key']) ? $p['key'] : "";
```
Expected: no issues emitted.

Same code in global namespace works fine.

**Comment by weirdan** [comment]
Possibly related to #619

**Comment by muglug** [comment]
Definitely related. Will fix tonight.
